File: Criacional/Builder/Builder01.py
class Motor:
    def __init__(self, tipo):
        self.tipo = tipo


class Roda:
    def __init__(self, tamanho):
        self.tamanho = tamanho


class Interior:
    def __init__(self, cor):
        self.cor = cor


class Carroceria:
    def __init__(self, estilo):
        self.estilo = estilo


class CarroBuilder:
    def __init__(self):
        self.motor = None
        self.roda = None
        self.interior = None
        self.carroceria = None

    def addMotor(self, tipo):
        self.motor = Motor(tipo)
        return self

    def addRoda(self, tamanho):
        self.roda = Roda(tamanho)
        return self

    def addInterior(self, cor):
        self.interior = Interior(cor)
        return self

    def addCarroceria(self, estilo):
        self.carroceria = Carroceria(estilo)
        return self

    def construir(self):
        return Carro(self.motor, self.carroceria, self.roda, self.interior)


class Carro:
    def __init__(self, motor, carroceria, rodas, interior):
        self.motor = motor
        self.carroceria = carroceria
        self.rodas = rodas
        self.interior = interior

    def mostrarDetalhes(self):
        print(f"""
Carro composto por:
Motor: {self.motor.tipo}
Carroceria: {self.carroceria.estilo}
Roda: {self.rodas.tamanho}
Interior: {self.interior.cor}
""")


builder = CarroBuilder()

carro1 = builder \
    .addMotor("1.0") \
    .addCarroceria("Pickup") \
    .addRoda(15) \
    .addInterior("Rosa") \
    .construir()

carro2 = builder \
    .addMotor("2.0") \
    .addCarroceria("SUV") \
    .addRoda(18) \
    .addInterior("Preto") \
    .construir()

carro1.mostrarDetalhes()
carro2.mostrarDetalhes()

File: Criacional/Builder/Builder01.js
//Define as partes de um Carro
class Motor{
    constructor(tipo){
        this.tipo = tipo;
    }
}

class Carroceria{
    constructor(estilo){
        this.estilo = estilo;
    }
}

class Rodas{
    constructor(tamanho){
        this.tamanho = tamanho;
    }
}

class Interior{
    constructor(cor){
        this.cor = cor;
    }
}

//Builder
class CarroBuilder{
    constructor(){
        this.motor = null;
        this.carroceria = null;
        this.rodas = null;
        this.interior = null
    }

    addMotor(tipo){
        this.motor = new Motor(tipo);
        return this;
    }

    addCarroceria(estilo){
        this.carroceria = new Carroceria(estilo);
        return this;
    }

    addRodas(tamanho){
        this.rodas = new Rodas(tamanho);
        return this;
    }

    addInterior(cor){
        this.interior = new Interior(cor);
        return this;
    }

    construir(){
        return new Carro(this.motor, this.carroceria, this.rodas, this.interior);
    }
}

//Construir Carro
class Carro{
    constructor(motor, carroceria, rodas, interior){
        this.motor = motor;
        this.carroceria = carroceria;
        this.rodas = rodas;
        this.interior = interior;
    }

    mostrarDetalhes(){
        console.log(`Carro composto por: \n
            Motor: ${this.motor.tipo} \n
            Carroceria: ${this.carroceria.estilo} \n
            Rodas: ${this.rodas.tamanho} \n
            Interior: ${this.interior.cor} \n`)
    }
}


//Cliente - Usando o Builder para construir o carro
const builder = new CarroBuilder();

const carro1 = builder
  .addMotor("1.0")
  .addCarroceria("Pickup")
  .addRodas(15)
  .addInterior("Rosa")
  .construir();

const carro2 = builder
  .addMotor("2.0")
  .addCarroceria("SUV")
  .addRodas(18)
  .addInterior("Preto")
  .construir();

carro1.mostrarDetalhes();
carro2.mostrarDetalhes();


    